<a href="https://colab.research.google.com/github/Befezdow/ai_test/blob/master/lab4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import drive
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import random
%matplotlib inline

drive.mount('/content/gdrive')
names_array = ['gender', 'age', 'class1', 'class2', 'class3', *['t{}'.format(i + 1) for i in range(60000)]]
url = '/content/gdrive/My Drive/ML/half.csv'
data = pd.read_csv(url, names=names_array)

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


Подготавливаем данные, заполняя пропущенные значения медианой.

In [2]:
data = data.fillna(data.median(axis = 0), axis = 0)
data

gender  age  class1  class2  ...    t59997    t59998    t59999    t60000
0     Female   85       3     5.0  ... -0.046202 -0.042003 -0.034466 -0.039000
1       Male   79       9     5.0  ... -0.797862 -0.819597 -0.771195 -0.845490
2       Male   66       3     5.0  ...  0.027947  0.030947  0.034947  0.037947
3       Male   93       2     5.0  ... -0.011236 -0.016540 -0.016289 -0.022542
4       Male   65       2     5.0  ... -0.031630 -0.026058 -0.012157 -0.001676
...      ...  ...     ...     ...  ...       ...       ...       ...       ...
1245    Male   74       2     5.0  ...  0.677640  0.675640  0.669640  0.661640
1246  Female   85       7     5.0  ... -0.053209 -0.053092 -0.049083 -0.049954
1247    Male   61       2     5.0  ...  0.064192  0.072895  0.062235  0.060652
1248  Female   85       3     5.0  ... -0.034080 -0.032233 -0.030831 -0.033880
1249    Male   82       1     5.0  ...  0.311619 -0.035582 -0.215320 -0.406685

[1250 rows x 60005 columns]

Осуществляем аугментацию данных. На каждый датчик идет по 5000 записей. Разбиваем так, чтобы на каждый приходилось по 250, тем самым размножая данные.

In [3]:
count_of_iterations = 20
abductions_count = 12
needed_per_abduction = 250

non_sliceable_columns = data.iloc[:,0:5]
sliceable_columns = data.iloc[:,5:]
sliceable_columns_count = sliceable_columns.shape[1]
per_abduction = int(sliceable_columns_count / abductions_count)

new_data = pd.DataFrame()
for i in range(count_of_iterations):
  random_position = random.randint(0, per_abduction - needed_per_abduction)

  total_columns_slice = pd.DataFrame()
  for j in range(abductions_count):
    start_column = j * per_abduction + random_position
    end_column = j * per_abduction + random_position + needed_per_abduction
    columns_slice = sliceable_columns.iloc[:, start_column : end_column]
    total_columns_slice = pd.concat([total_columns_slice, columns_slice.copy()], axis=1, sort=False, ignore_index=True)
  
  completed_slice = pd.concat([non_sliceable_columns.copy(), total_columns_slice], axis=1, sort=False, ignore_index=True)
  new_data = pd.concat([new_data, completed_slice], axis=0, sort=False, ignore_index=True)

new_data.columns = ['gender', 'age', 'class1', 'class2', 'class3', *['t{}_{}'.format(i + 1, j + 1) for i in range(abductions_count) for j in range(needed_per_abduction)]]
new_data

gender  age  class1  class2  ...   t12_247   t12_248   t12_249   t12_250
0      Female   85       3     5.0  ...  0.238549  0.245322  0.251146  0.256163
1        Male   79       9     5.0  ... -0.097346 -0.102382 -0.107586 -0.112518
2        Male   66       3     5.0  ... -0.052053 -0.051053 -0.051053 -0.051053
3        Male   93       2     5.0  ... -0.099614 -0.096544 -0.093858 -0.092545
4        Male   65       2     5.0  ... -0.037603 -0.041808 -0.045062 -0.041955
...       ...  ...     ...     ...  ...       ...       ...       ...       ...
24995    Male   74       2     5.0  ... -0.397360 -0.388360 -0.377360 -0.368360
24996  Female   85       7     5.0  ... -0.112335 -0.107062 -0.107498 -0.101975
24997    Male   61       2     5.0  ...  0.159102  0.160464  0.167530  0.165822
24998  Female   85       3     5.0  ... -0.016949 -0.016567 -0.016735 -0.018542
24999    Male   82       1     5.0  ... -0.019741 -0.016817 -0.012379 -0.007662

[25000 rows x 3005 columns]

In [0]:
data = new_data

Выкидываем лишние классовые признаки, а затем осуществляем векторизацию и нормализацию данных.

In [5]:
# убираем лишние классовые признаки
data = data.drop('class2', axis = 1)
data = data.drop('class3', axis = 1)
 
#собираем числовые колонки
numerical_columns = [c for c in data.columns if data[c].dtype.name != 'object' and c != 'class1']

# векторизируем колонку gender
data.at[data['gender'] == 'Male', 'gender'] = 0
data.at[data['gender'] == 'Female', 'gender'] = 1

# нормализуем числовые атрибуты
data_numerical = data[numerical_columns]
data_numerical = (data_numerical - data_numerical.mean(axis = 0))/data_numerical.std(axis = 0)
data[numerical_columns] = data_numerical

data

gender       age  class1  ...   t12_248   t12_249   t12_250
0           1  1.177837       3  ...  0.838785  0.860384  0.879626
1           0  0.808686       9  ... -0.362135 -0.381297 -0.399656
2           0  0.008860       3  ... -0.184852 -0.185620 -0.186379
3           0  1.670038       2  ... -0.341973 -0.333781 -0.330350
4           0 -0.052666       2  ... -0.152921 -0.164885 -0.154810
...       ...       ...     ...  ...       ...       ...       ...
24995       0  0.501061       2  ... -1.349865 -1.315069 -1.287396
24996       1  1.177837       7  ... -0.378299 -0.380994 -0.363071
24997       0 -0.298766       2  ...  0.545698  0.570961  0.566150
24998       1  1.177837       3  ... -0.065744 -0.066836 -0.073568
24999       0  0.993262       1  ... -0.066606 -0.051759 -0.035817

[25000 rows x 3003 columns]

Разбиваем данные на входы и выходы.

In [0]:
# бьем данные на входы и выходы
x = data.drop('class1', axis = 1)
y = data['class1']

Разбиваем данные на обучающую и тестовую выборки.

In [7]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3, random_state = 42)

n_train, _ = x_train.shape 
n_test,  _ = x_test.shape 

print('Train dataset size: {}'.format(n_train))
print('Test dataset size: {}'.format(n_test))

Train dataset size: 17500
Test dataset size: 7500


# KMeans

In [0]:
from sklearn.cluster import KMeans

kmeans = KMeans(n_clusters=len(new_data['class1'].unique()), random_state=42).fit(x)

In [9]:
kmeans_cluster = kmeans.predict(x) + 1
print(np.unique(kmeans.labels_))
check = kmeans_cluster == y
np.where(check == True)[0].shape

[0 1 2 3 4 5 6 7 8]


(1522,)

Количество классов совпадает, т.к. модель подразумевает передачу их количества. Однако стоит заметить, что пересечение с реальным разбиением очень слабое (1522 из 25000).

# DBSCAN

In [10]:
from sklearn.cluster import DBSCAN

dbscan = DBSCAN().fit(x)
np.unique(dbscan.labels_)

array([-1])

DBSCAN дает разбиение всего на 1 класс, что является не приемлемым, так как изначальные данные бьются на 9 классов.

# AffinityPropagation

In [1]:
from sklearn.cluster import AffinityPropagation

affinity = AffinityPropagation().fit(x)
np.unique(affinity.labels_)

NameError: ignored

Не получилось проверить данную модель, так как она не помещается в оперативную память (более 13 гб).

# Spectral Clustering
 

In [0]:
from sklearn.cluster import SpectralClustering

spectral = SpectralClustering(n_clusters=len(new_data['class1'].unique()), random_state=42).fit(x)
spectral_cluster = spectral.predict(x) + 1
print(np.unique(spectral.labels_))
check_spectral = spectral_cluster == y
np.where(check_spectral == True)[0].shape

Также не получилось проверить данную модель, так как она не помещается в оперативную память (более 13 гб).

# Mean-shift

In [11]:
from sklearn.cluster import MeanShift, estimate_bandwidth

bandwidth = estimate_bandwidth(x, quantile=0.2, n_samples=1000)

ms = MeanShift(bandwidth=bandwidth, bin_seeding=True).fit(x)
ms_cluster = ms.predict(x) + 1
print(np.unique(ms.labels_))
check_ms = ms_cluster == y
np.where(check_ms == True)[0].shape

[0]


(3340,)

Mean-shift, также как и DBSCAN, дает разбиение всего на 1 класс, что является не приемлемым, так как изначальные данные бьются на 9 классов. Пересечение с реальным разбиением также слабое (3340 из 25000).

Исходя из результатов, можно сделать вывод что данные плохо поддаются кластеризации. Модели, которые не запрашивают кол-во кластеров, не могут разделить данные и в итоге получается всего 1 кластер. KMeans же принимает кол-во классов, но дает очень плохую точность (менее 20%).